src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.contoso.com") : null; } }
+            };
+        }
+
+        // These builders exhaustively populate the generated SARIF file, including non-required JSON primitives
+        public static IDictionary<Type, PrimitiveValueBuilder> PopulateAllPrimitives()
+        {
+            // The values in this output are designed to allow conformance to the schema. In our schema,.
+            // for example, some integer values must be > 0. And so, out optional integer value is 
+            // written as 1.
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return true; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 1; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 1; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "[Required string value]" : "[Optional string value]"; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(1776, 7, 4).ToUniversalTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.value.contoso.com") : new Uri("https://optional.uri.value.contoso.com"); } }
+            };
+        }
+
+        public DefaultObjectPopulatingVisitor(JsonSchema schema, IDictionary<Type, PrimitiveValueBuilder> primitiveValueBuilders = null)
+        {
+            _schema = schema;
+            _typeToPropertyValueConstructorMap = primitiveValueBuilders ?? PopulateRequiredPrimitives();
+        }
+
+        private JsonSchema _schema;
+        private IDictionary<Type, PrimitiveValueBuilder> _typeToPropertyValueConstructorMap;
+
+        // There are two places in the format where object nest references to the defining 
+        // type. 'exception.innerExceptions' is an array of exception objects. 
+        // 'node.children' is another collection of nodes. We need to watch for these
+        // scenarios to prevent re-entrancy and eventual stack consumption when 
+        // producing samples of these types.
+        private int graphNodeParentCount;
+        private int exceptionDataParentCount;
+
+        public override object VisitActual(ISarifNode node)
+        {
+            PopulateInstanceWithDefaultMemberValues(node);
+
+            return base.VisitActual(node);
+        }
+
+        // Retain nesting level for visiting exceptions to prevent
+        // unbounded re-entrance populating exception.innerExceptions
+        public override ExceptionData VisitExceptionData(ExceptionData node)
+        {
+            exceptionDataParentCount++;
```

Comment:
>exceptionDataParentCount [](start = 12, length = 24)

While you're counting, do you want to throw if the nesting level exceeds a maximum?

Reply:
Nesting is impossible due to other code (that refuses to populate data members to recurse into).

---
In reply to: [229110609](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229110609) [](ancestors = 229110609)